packages/yoastseo/src/parse/build/private/getSentencePositions.js

```diff
@@ -0,0 +1,120 @@
+ */
+function adjustSentenceEnd( descendantNodes, descendantTagPositions, sentenceStart, sentenceEnd ) {
+	/*
+	 * If the start position of a descendant's node tag is between the start and end position of the sentence, or is
+	 * the same as the start/end position of the sentence, add the tag's length to the end position of the sentence.
+	 */
+	descendantTagPositions.forEach( ( position ) => {
+		if ( position.startOffset >= sentenceStart && position.startOffset <= sentenceEnd ) {
+			sentenceEnd += ( position.endOffset - position.startOffset );
+		}
+	} );
+	/*
+	 * If the length of a start tag at the end of the sentence was added to the sentenceEnd in the step above,
+	 * remove it. We are using the data from the original descendantNodes array for this check, as the
+	 * descendantTagPositions array doesn't specify whether each tag is an opening or closing tag.
+	 */
+	const startTagAtEndOfSentence = descendantNodes.find( node =>
+		node.sourceCodeLocation.startTag.endOffset === sentenceEnd );
+	if ( startTagAtEndOfSentence ) {
+		const startTagLocation = startTagAtEndOfSentence.sourceCodeLocation.startTag;
+		sentenceEnd = sentenceEnd - ( startTagLocation.endOffset - startTagLocation.startOffset );
+	}
+
+	return sentenceEnd;
+}
+
+/**
+ * Gets the start and end positions of the sentences in the source code.
+ *
+ * @param {Node} 		node  		The paragraph or heading node.
+ * @param {Sentence[]} 	sentences 	The sentences in the node.
+ *
+ * @returns {Sentence[]} The sentences, with their positions in the source code.
+ */
+export default function getSentencePositions( node, sentences ) {
+	/*
+	 * We cannot calculate sentence positions if there are no sentences, or if we don't know the node's source code
+	 * location.
+	 */
+	if (  sentences.length === 0 || ! node.sourceCodeLocation ) {
+		return sentences;
+	}
+	/*
+	 * Set the start position of the first sentence. If the node is an implicit paragraph, which don't have start and
+	 * end tags, set the start position to the start of the node. Otherwise, set the start position to the end of the
+	 * node's start tag.
+	 */
+	let sentenceStart = node.name === "p" && node.isImplicit
+		? node.sourceCodeLocation.startOffset
+		: node.sourceCodeLocation.startTag.endOffset;
+	let sentenceEnd;
+	let descendantNodes = [];
+	let descendantTagPositions = [];
+
+	/*
+	 * Check if the node has any descendant nodes that have a sourceCodeLocation property (all nodes other than Text nodes
+	 * should have this property). If such nodes exist, store the positions of each node's opening and closing tags in
+	 * an array. These positions will have to be taken into account when calculating the position of the sentences.
+	 */
+	descendantNodes = node.findAll( descendantNode => descendantNode.sourceCodeLocation );
+	if ( descendantNodes.length > 0 ) {
+		descendantTagPositions = getDescendantPositions( descendantNodes );
+	}
+
+	for ( let i = 0; i < sentences.length; i++ ) {
```

Comment:
I think you can use `sentences.forEach` here as well, makes the code a little bit more readable.